src/test/java/build/buildfarm/instance/memory/MemoryInstanceTest.java

```diff
@@ -158,4 +174,103 @@ public void actionCacheRetrievableByActionKey() {
     assertThat(instance.getActionResult(
         instance.getDigestUtil().computeActionKey(action))).isEqualTo(result);
   }
+
+  @Test
+  public void missingOperationWatchInvertsWatchInitialState() {
+    Predicate<Operation> watcher = (Predicate<Operation>) mock(Predicate.class);
+    // a request for a watch on an operation that does not exist must
+    // invert watchInitialState to indicate that it will not call the watcher
+    // again if it has not handled the completed state
+    assertThat(instance.watchOperation(
+        "does-not-exist",
+        /* watchInitialState=*/ false,
+        watcher)).isTrue();
+    verifyZeroInteractions(watcher);
+
+    Predicate<Operation> watchInitialWatcher = (Predicate<Operation>) mock(Predicate.class);
+    when(watchInitialWatcher.test(eq(null))).thenReturn(true);
+    assertThat(instance.watchOperation(
+        "does-not-exist",
+        /* watchInitialState=*/ true,
+        watchInitialWatcher)).isFalse();
+    verify(watchInitialWatcher, times(1)).test(eq(null));
+  }
+
+  @Test
+  public void initialWatchWithCompletedSignalsWatching() {
+    Predicate<Operation> watcher = (Predicate<Operation>) mock(Predicate.class);
+    when(watcher.test(eq(null))).thenReturn(false);
+    assertThat(instance.watchOperation(
+        "does-not-exist",
+        /* watchInitialState=*/ true,
+        watcher)).isTrue();
+    verify(watcher, times(1)).test(eq(null));
+  }
+
+  @Test
+  public void watchOperationAddsWatcher() {
+    Operation operation = Operation.newBuilder()
+        .setName("my-watched-operation")
+        .build();
+    outstandingOperations.put(operation.getName(), operation);
+
+    List<Predicate<Operation>> operationWatchers =
+        (List<Predicate<Operation>>) mock(List.class);
+    watchers.put(operation.getName(), operationWatchers);
+
+    Predicate<Operation> watcher = (Predicate<Operation>) mock(Predicate.class);
+    assertThat(instance.watchOperation(
+        operation.getName(),
+        /* watchInitialState=*/ false,
+        watcher)).isTrue();
+    verify(operationWatchers, times(1)).add(eq(watcher));
```

Comment:
Style: prefer not mocking data objects. In this case, you could use an actual list and verify that it .containsExactly(yourItems)